### PR TITLE
"PROMOTE <battleid>" support

### PR DESCRIPTION
**Pull request by MasterBel2**
I've added support for a command "PROMOTE <battleid>" which when recieved by the server will be distributed to all other clients. If this is used instead of promoting in channels it will allow for all host bots to be removed from #main

### CODE REVIEW

**Comment by abma** [comment]
in short: horrible, can't merge.

- contains debug print
- contains wrong whitespace
- broadcast_Promote() isn't needed
- JSON should be used (thats basicly new) to make changes / new parameters more easy


thoughts about this?
https://github.com/spring/uberserver/commit/84f0881830e2284cea723b3cef85a4a5d0c902c0

**Comment by MasterBel2** [comment]
Oops, I was sure I deleted the debug prints. I'll make sure to double-check in future.
On whitespace: not sure what you mean. Please enlighten me?

File: DataHandler.py [hunk 1]
```diff
@@ -128,7 +128,7 @@ def _fk_pragma_on_connect(dbapi_con, con_record):
 			self.channels[name] = newchan
 
 		self.parseFiles()
-		self.protocol = Protocol.Protocol(self)
+		self.protocol = Protocol.Protocol(self)
```
Comment (abma) [inline DataHandler.py]: whitespace was added here...

Comment (abma) [inline DataHandler.py]: whitespace was added here...

File: protocol/Protocol.py [hunk 2]
```diff
@@ -637,6 +638,16 @@ def broadcast_SendBattle(self, battle, data, sourceClient=None):
 			client = self.clientFromSession(session_id)
 			if sourceClient == None or not sourceClient.db_id in client.ignored:
 				client.Send(data)
+				
+	def broadcast_Promote(self, promotedBattle, client):
+		for name, reciever in self._root.usernames.items():
+			if client.session_id == receiver.session_id: # don't send PROMOTE to self
+				print('not sending promote to self')
```
Comment (abma) [inline protocol/Protocol.py]: debug print

File: protocol/Protocol.py [hunk 3]
```diff
@@ -1599,6 +1610,15 @@ def in_OPENBATTLE(self, client, type, natType, password, port, maxplayers, hashc
 		client.Send('JOINBATTLE %s %s' % (battle_id, hashcode))
 		client.Send('REQUESTBATTLESTATUS')
 
+	def in_PROMOTE(self, client):
+		print('Promote command recieved')
```
Comment (abma) [inline protocol/Protocol.py]: debug print

File: protocol/Protocol.py [hunk 4]
```diff
@@ -1599,6 +1610,15 @@ def in_OPENBATTLE(self, client, type, natType, password, port, maxplayers, hashc
...
+		promoted_battle = client.current_battle
+		if not promoted_battle:
+			print('client is not in a battle')
+			return
+		print('client is in a battle')
```
Comment (abma) [inline protocol/Protocol.py]: debug print

File: protocol/Protocol.py [hunk 5]
```diff
@@ -1599,6 +1610,15 @@ def in_OPENBATTLE(self, client, type, natType, password, port, maxplayers, hashc
 		client.Send('JOINBATTLE %s %s' % (battle_id, hashcode))
 		client.Send('REQUESTBATTLESTATUS')
 
+	def in_PROMOTE(self, client):
+		print('Promote command recieved')
+		promoted_battle = client.current_battle
+		if not promoted_battle:
+			print('client is not in a battle')
```
Comment (abma) [inline protocol/Protocol.py]: debug print

File: protocol/Protocol.py [hunk 6]
```diff
@@ -637,6 +638,16 @@ def broadcast_SendBattle(self, battle, data, sourceClient=None):
 			client = self.clientFromSession(session_id)
 			if sourceClient == None or not sourceClient.db_id in client.ignored:
 				client.Send(data)
+				
+	def broadcast_Promote(self, promotedBattle, client):
```
Comment (abma) [inline protocol/Protocol.py]: not needed function, self._root.broadcast should be used instead.

**Comment by MasterBel2** [comment]
JSON for new commands defnitely makes sense. 
And thanks very much for the other feedback, it all makes sense now.

**Comment by abma** [comment]
FYI: the current development version is used on the lobby server

**Pull request closed without merging**